Sources/FoundationEssentials/FileManager/FileOperations.swift

```diff
@@ -342,52 +348,69 @@ enum _FileOperations {
-                            }
-                            guard filemanager?._shouldRemoveItemAtPath(entry.fileNameWithPrefix) ?? true else { return }
-                            if entry.dwFileAttributes & FILE_ATTRIBUTE_DIRECTORY == FILE_ATTRIBUTE_DIRECTORY {
-                                if !RemoveDirectoryW($0) {
-                                    let error = CocoaError.removeFileError(GetLastError(), entry.fileName)
-                                    guard (filemanager?._shouldProceedAfter(error: error, removingItemAtPath: entry.fileNameWithPrefix) ?? false) else {
-                                        throw error
+                            } else {
+                                if entry.dwFileAttributes & FILE_ATTRIBUTE_READONLY == FILE_ATTRIBUTE_READONLY {
+                                    guard SetFileAttributesW($0, entry.dwFileAttributes & ~FILE_ATTRIBUTE_READONLY) else {
+                                        throw CocoaError.removeFileError(GetLastError(), path)
                                     }
                                 }
-                            } else {
-                                if !DeleteFileW($0) {
-                                    let error = CocoaError.removeFileError(GetLastError(), entry.fileName)
-                                    guard (filemanager?._shouldProceedAfter(error: error, removingItemAtPath: entry.fileNameWithPrefix) ?? false) else {
-                                        throw error
+                                if entry.dwFileAttributes & FILE_ATTRIBUTE_DIRECTORY == FILE_ATTRIBUTE_DIRECTORY {
+                                    guard filemanager?._shouldRemoveItemAtPath(path) ?? true else { return }
+                                    if !RemoveDirectoryW($0) {
+                                        let error = CocoaError.removeFileError(GetLastError(), entry.fileName)
+                                        guard (filemanager?._shouldProceedAfter(error: error, removingItemAtPath: entry.fileNameWithPrefix) ?? false) else {
+                                            throw error
+                                        }
+                                    }
+                                } else {
+                                    guard filemanager?._shouldRemoveItemAtPath(path) ?? true else { return }
+                                    if !DeleteFileW($0) {
+                                        let error = CocoaError.removeFileError(GetLastError(), entry.fileName)
+                                        guard (filemanager?._shouldProceedAfter(error: error, removingItemAtPath: entry.fileNameWithPrefix) ?? false) else {
+                                            throw error
+                                        }
                                     }
                                 }
                             }
                         }
                     }
+
+                    if RemoveDirectoryW($0) { return }
```

Comment:
Which directory does this call remove? Since we put the original `path` on the stack, shouldn't iterating the stack take care of everything?

Reply:
This is related to the previous point - we need to do this to ensure that we do not prompt twice for the same directory. The Unix side takes care of this through FTS having a post-order traversal callback.